ksonnet-gen/ksonnet/emitter.go

```diff
@@ -0,0 +1,469 @@
+package ksonnet
+
+import (
+	"fmt"
+	"log"
+	"sort"
+	"strings"
+
+	"github.com/ksonnet/ksonnet-lib/ksonnet-gen/kubespec"
+)
+
+// Emit takes a swagger API specification, and returns the text of
+// `ksonnet-lib`, written in Jsonnet.
+func Emit(spec *kubespec.APISpec) ([]byte, error) {
+	root := newEmitter(spec)
+
+	m := newMarshaller()
+	return root.emit(m)
+}
+
+//-----------------------------------------------------------------------------
+// Emitter.
+//-----------------------------------------------------------------------------
+
+// `root` is an abstraction of the root of `k8s.libsonnet`, which can be
+// emitted as Jsonnet code using the `emit` method.
+//
+// `root` contains and manages a set of `groups`, which represent a
+// set of Kubernetes API groups (e.g., core, apps, extensions), and
+// holds all of the logic required to build the `groups` from an
+// `kubespec.APISpec`.
+type root struct {
+	info   *kubespec.SchemaInfo
+	groups groupSet // set of groups, e.g., core, apps, extensions.
+}
+
+func newEmitter(spec *kubespec.APISpec) *root {
+	root := root{
+		info:   spec.Info,
+		groups: make(groupSet),
+	}
+
+	for defName, def := range spec.Definitions {
+		root.addDefinition(defName, def)
+	}
+
+	return &root
+}
+
+func (root *root) emit(m *marshaller) ([]byte, error) {
+	m.bufferLine("// AUTOGENERATED from the Kubernetes OpenAPI specification. DO NOT MODIFY.")
```

Comment:
Hmm, on second thought, we need to think through the provenance story here. Hashing `swagger.json` will only tell you if you generated your spec from the same data. Probably what we want to do is to include the HEAD of the Kubernetes repository, so that it's fully reproducible.